src/NHibernate.Test/NHSpecificTest/GH1921/Fixture.cs

```diff
@@ -0,0 +1,85 @@
+﻿using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH1921
+{
+	[TestFixture]
+	public class Fixture : BugTestCase
+	{
+		protected override void OnSetUp()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				var e1 = new Entity {Name = "Bob"};
+				session.Save(e1);
+
+				var e2 = new Entity {Name = "Sally"};
+				session.Save(e2);
+
+				transaction.Commit();
+			}
+		}
+
+		protected override void OnTearDown()
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				session.CreateQuery("delete from System.Object").ExecuteUpdate();
+
+				transaction.Commit();
+			}
+		}
+
+		[Theory]
+		public void DmlInsert(bool filtered)
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				if (filtered)
+					session.EnableFilter("NameFilter").SetParameter("name", "Bob");
+				var rowCount = session.CreateQuery("insert into Entity (Name) select e.Name from Entity e").ExecuteUpdate();
+				transaction.Commit();
+
+				// If the DML has to take the filter into account, then below Assert should be
+				// Assert.That(rowCount, Is.EqualTo(filtered ? 1 : 2));
+				Assert.That(rowCount, Is.EqualTo(2));
+			}
+		}
+
+		[Theory]
+		public void DmlUpdate(bool filtered)
+		{
+			using (var session = OpenSession())
+			using (var transaction = session.BeginTransaction())
+			{
+				if (filtered)
+					session.EnableFilter("NameFilter").SetParameter("name", "Bob");
+				var rowCount = session.CreateQuery("update Entity e set Name = 'newName'").ExecuteUpdate();
+				transaction.Commit();
+
+				// If the DML has to take the filter into account, then below Assert should be
+				// Assert.That(rowCount, Is.EqualTo(filtered ? 1 : 2));
+				Assert.That(rowCount, Is.EqualTo(2));
```

Comment:
Currently, update and delete do not care at all about filters enabled in the session. They are ignored.
So likely the fix for the insert case should be to ignore them completely too.

Otherwise, if the insert case is fixed by actually supporting filters, then the update and delete would need to support it too. And it would be a breaking change for those already using an update or delete with a filter enabled: their update or delete would change by taking into account the filter.